### PR TITLE
server: fix diskoffering details in vm response

### LINKED ISSUE #8120: ACS 4.18 diskofferingid missing from deployVirtualMachine response
##### ISSUE TYPE
 * Bug Report

##### COMPONENT NAME
~~~
API
~~~

##### CLOUDSTACK VERSION
~~~
4.18
~~~

##### CONFIGURATION
Hypervisor: XenServer

##### OS / ENVIRONMENT
CentOS 7

##### SUMMARY
In ACS up until 4.11.2.0, the response to the _deployVirtualMachine_ API call contained the data disk _diskofferingid_ and _diskofferingname_

I noticed that on ACS 4.18, this is no longer the case.

##### STEPS TO REPRODUCE
deploy a virtual machine with disk offering and size set.

API call extracted from ACS log file:
~~~
2023-10-19 09:41:56,493 DEBUG [c.c.a.ApiServlet] (qtp1444635922-841947:ctx-fcdbc879) (logid:08db5be6) ===START===  10.12.0.248 -- GET  details%5B0%5D.cpuNumber=2&serviceOfferingId=8b6add88-0261-4d05-b04b-146032b8f35b&networkids=6710de0c-d4f5-48bf-bd8b-1543c4d3f888&size=20&details%5B0%5D.cpuSpeed=1000&displayname=test-deb-4%20test-deb-4&name=test-deb-4&zoneId=e99e796c-4e5c-4549-a0d7-ddd3d8526c06&templateId=e9606f2c-a1d1-4548-ae52-00dc36e17662&command=deployVirtualMachine&details%5B0%5D.memory=2048&diskofferingid=05255aa1-1ef4-4698-a151-67809848a646&apiKey=xxxx&signature=xxxx
~~~

##### EXPECTED RESULTS

My application polls the job result, and also depends on the returned diskofferingid and diskofferingname fields in the resulting XML document. But both fields are missing, making the application think there is no data disk.

Before the upgrade to ACS 4.18, this did work.
The API docs still list both fields for the response: https://cloudstack.apache.org/api/apidocs-4.18/apis/deployVirtualMachine.html

##### ACTUAL RESULTS

The fields are missing now. Result of the job: 

~~~
<?xml version="1.0" encoding="UTF-8"?>
<queryasyncjobresultresponse cloud-stack-version="4.18.1.0-SNAPSHOT">
    <accountid>813308c2-8405-46c3-aa09-0e6bbd90e3b1</accountid>
    <userid>5f52698d-cdd1-4ce2-ad15-128e414eb42d</userid>
    <cmd>org.apache.cloudstack.api.command.user.vm.DeployVMCmd</cmd>
    <jobstatus>1</jobstatus>
    <jobprocstatus>0</jobprocstatus>
    <jobresultcode>0</jobresultcode>
    <jobresulttype>object</jobresulttype>
    <jobresult>
        <virtualmachine>
            <id>c4563c04-4e2f-43df-874f-44b90c6627a4</id>
            <name>test-deb-4</name>
            <displayname>test-deb-4 test-deb-4</displayname>
            <account>admin</account>
            <userid>5f52698d-cdd1-4ce2-ad15-128e414eb42d</userid>
            <username>admin</username>
            <domainid>f8082002-79f2-41f8-affe-7b8328a8dd20</domainid>
            <domain>XX</domain>
            <created>2023-10-19T09:41:56+0200</created>
            <lastupdated>2023-10-19T09:42:33+0200</lastupdated>
            <state>Running</state>
            <haenable>true</haenable>
            <zoneid>e99e796c-4e5c-4549-a0d7-ddd3d8526c06</zoneid>
            <zonename>XXX</zonename>
            <hostcontrolstate>Enabled</hostcontrolstate>
            <templateid>e9606f2c-a1d1-4548-ae52-00dc36e17662</templateid>
            <templatename>debian-12</templatename>
            <templatedisplaytext>Debian 12.2 20231018</templatedisplaytext>
            <passwordenabled>false</passwordenabled>
            <serviceofferingid>8b6add88-0261-4d05-b04b-146032b8f35b</serviceofferingid>
            <serviceofferingname>xxx</serviceofferingname>
            <cpunumber>2</cpunumber>
            <cpuspeed>1000</cpuspeed>
            <memory>2048</memory>
            <guestosid>9a37c197-7197-4eee-93ef-e428560d4244</guestosid>
            <rootdeviceid>0</rootdeviceid>
            <rootdevicetype>ROOT</rootdevicetype>
            <nic>
                <id>903ff7fa-c03d-4277-b6e6-c86216090921</id>
                <networkid>6710de0c-d4f5-48bf-bd8b-1543c4d3f888</networkid>
                <networkname>xxx</networkname>
                <netmask>xxx</netmask>
                <gateway>xxx</gateway>
                <ipaddress>xxx</ipaddress>
                <isolationuri>xxx</isolationuri>
                <broadcasturi>xxx</broadcasturi>
                <traffictype>Guest</traffictype>
                <type>Shared</type>
                <isdefault>true</isdefault>
                <macaddress>xxx</macaddress>
                <deviceid>0</deviceid>
            </nic>
            <hypervisor>XenServer</hypervisor>
            <details>
                {cpuNumber=2, memory=2048, cpuSpeed=1000, hypervisortoolsversion=xenserver56,
                platform=device-model:qemu-upstream-compat;apic:true;viridian:true;timeoffset:0;pae:true;acpi:1;hpet:true;secureboot:false;nx:true}</details>
            <readonlydetails>dataDiskController, rootDiskController</readonlydetails>
            <isdynamicallyscalable>true</isdynamicallyscalable>
            <ostypeid>9a37c197-7197-4eee-93ef-e428560d4244</ostypeid>
            <osdisplayname>Debian GNU/Linux 12 (64-bit)</osdisplayname>
            <pooltype>PreSetup</pooltype>
            <receivedbytes>0</receivedbytes>
            <sentbytes>0</sentbytes>
            <hasannotations>false</hasannotations>
            <jobid>abbf6274-ccd1-437d-912a-d94999e8771c</jobid>
            <jobstatus>0</jobstatus>
        </virtualmachine>
    </jobresult>
    <jobinstancetype>VirtualMachine</jobinstancetype>
    <jobinstanceid>c4563c04-4e2f-43df-874f-44b90c6627a4</jobinstanceid>
    <created>2023-10-19T09:41:56+0200</created>
    <completed>2023-10-19T09:42:33+0200</completed>
    <jobid>abbf6274-ccd1-437d-912a-d94999e8771c</jobid>
</queryasyncjobresultresponse>
~~~

### PR DESCRIPTION
**Pull request by shwstppr**
### Description

Fixes #8120

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)
- [ ] build/CI

#### Feature/Enhancement Scale

- [ ] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [ ] Major
- [ ] Minor
- [ ] Trivial

### CODE REVIEW

**Comment by shwstppr** [comment]
Not sure if we should return DATA disk offering details after #5008 but this should allow fixing the issue as we have multiple entries in user_vm_view when there is a data disk @harikrishna-patnala

**Comment by weizhouapache** [comment]
> Not sure if we should return DATA disk offering details after #5008 but this should allow fixing the issue as we have multiple entries in user_vm_view when there is a data disk @harikrishna-patnala

I have the same concern

**Comment by shwstppr** [comment]
@blueorangutan package

**Review by rohityadavcloud — APPROVED** [review]
LGTM - didn't test it

**Comment by shwstppr** [comment]
@blueorangutan test

**Review by DaanHoogland — APPROVED** [review]
clgtm

**Comment by shwstppr** [comment]
@harikrishna-patnala can you please have a look at this?

**Review by harikrishna-patnala — APPROVED** [review]
I've not changed this response in #5008. Not sure why is this missing.
But this PR change looks good to me.

**Comment by weizhouapache** [comment]
What if there are multiple data disks ?

**Comment by harikrishna-patnala** [comment]
> What if there are multiple data disks ?

for each disk there is a separate entry in user_vm_view. "testVM" has two disks and "test" has 3 disks
image

**Comment by weizhouapache** [comment]
> > What if there are multiple data disks ?
> 
> for each disk there is a separate entry in user_vm_view. "testVM" has two disks and "test" has 3 disks image

yes @harikrishna-patnala 
in this case, the diskofferingid and diskofferingname in the uservm response will be the uuid and name of the first data disk, right ?
is it expected result ?

**Comment by DaanHoogland** [comment]
@shwstppr is this not ready for review yet?

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test keepEnv

**Comment by shwstppr** [comment]
> @shwstppr is this not ready for review yet?

I'm not sure about this change. Although we seem to be returning diskofferingid of the first data disk even in the past but I think we can deprecate it or update its param description to say that it will be returned only in particular cases.
I'm okay to mark it ready if we have a consensus that the behaviour of returning first data disk offering id is fine

**Comment by DaanHoogland** [comment]
personally I think that if we include diskinfo in the VM response it should be an array of all disks.
If we don't have consensus, shall we move this to a newer release?

**Comment by DaanHoogland** [comment]
@shwstppr we are having second thoughts on how this should be implemented but in the mean while this is a regression to the user as reported in #8120. I think we should test and merge and considder how to improve later.
agree?
cc @harikrishna-patnala @weizhouapache @rohityadavcloud

**Comment by weizhouapache** [comment]
> @shwstppr we are having second thoughts on how this should be implemented but in the mean while this is a regression to the user as reported in #8120. I think we should test and merge and considder how to improve later. agree? cc @harikrishna-patnala @weizhouapache @rohityadavcloud

@DaanHoogland 
I would suggest to add all volumes info in the response, and ask users to get diskofferingid/name from volumes info, if they face the issue #8120

**Comment by weizhouapache** [comment]
> > > @shwstppr we are having second thoughts on how this should be implemented but in the mean while this is a regression to the user as reported in #8120. I think we should test and merge and considder how to improve later. agree? cc @harikrishna-patnala @weizhouapache @rohityadavcloud
> > 
> > 
> > @DaanHoogland I would suggest to add all volumes info in the response, and ask users to get diskofferingid/name from volumes info, if they face the issue #8120
> 
> I agree adding all to the volume list, but this is a backwards compatibility issue.

IMHO, we can give more information, rather than giving wrong info (if there are multiple volumes.
) just to keep backwards compatibility

**Comment by DaanHoogland** [comment]
> > > > @shwstppr we are having second thoughts on how this should be implemented but in the mean while this is a regression to the user as reported in #8120. I think we should test and merge and considder how to improve later. agree? cc @harikrishna-patnala @weizhouapache @rohityadavcloud
> > > 
> > > 
> > > @DaanHoogland I would suggest to add all volumes info in the response, and ask users to get diskofferingid/name from volumes info, if they face the issue #8120
> > 
> > 
> > I agree adding all to the volume list, but this is a backwards compatibility issue.
> 
> IMHO, we can give more information, rather than giving wrong info (if there are multiple volumes. ) just to keep backwards compatibility

I dissagree that the offering id and offering name of the root volume are wrong information.

In addition after analysing the DB and code I think it is not good to add the other information of the volumes to the VM response. There is a possibility to call `list volumes virtualmachineid=<uuid>` separately. If any bug needs solving it is that this one returns an empty offering and Id if you didn't choose one for the root volume.

**Comment by shwstppr** [comment]
@DaanHoogland @weizhouapache I suggest we go ahead with this PR for 4.18 branch and then in main create a PR,
- updating the params description that they are returned only when VM deployed from ISO or any other valid case.
- use listVirtualMachines API details param to return volume details

**Comment by weizhouapache** [comment]
> I dissagree that the offering id and offering name of the root volume are wrong information.

@DaanHoogland 
the `diskofferingid` and `diskofferingname` are the info of DATA disk, not ROOT disk.

**Comment by weizhouapache** [comment]
> @DaanHoogland @weizhouapache I suggest we go ahead with this PR for 4.18 branch and then in main create a PR,
> 
> * updating the params description that they are returned only when VM deployed from ISO or any other valid case.
> * use listVirtualMachines API details param to return volume details

@shwstppr 
I agree with the changes you mentioned.
AFAIK, `diskofferingid` and `diskofferingname` are the info of ROOT disk in 4.16.x. However, `disk_offering_id` has been removed from `vm_instance` table in #5008 , which leads to a bit complicated to set the value, but should be feasible.

that's say, this PR needs to be updated.

File: server/src/main/java/com/cloud/api/query/dao/UserVmJoinDaoImpl.java [hunk 1]
```diff
@@ -550,6 +552,11 @@ public UserVmResponse setUserVmResponse(ResponseView view, UserVmResponse userVm
             userVmData.addAffinityGroup(resp);
         }
 
+        if (StringUtils.isEmpty(userVmData.getDiskOfferingId()) && !Volume.Type.ROOT.equals(uvo.getVolumeType())) {
```
[the remark this answers is not in the document]
Comment (weizhouapache) [reply server/src/main/java/com/cloud/api/query/dao/UserVmJoinDaoImpl.java:555]: I think this should be added in `newUserVmResponse` method, not `setUserVmResponse`

[the remark this answers is not in the document]
Comment (shwstppr) [reply server/src/main/java/com/cloud/api/query/dao/UserVmJoinDaoImpl.java:555]: @weizhouapache for ROOT disk and ISO we are already setting the diskoffering. This PR change will only affect with VM deployed from templates having multiple disks.
In older version we were returning DATA disk's offering as disk offering

```
(localcloud) 🐱 > list managementservers 
{
  "count": 1,
  "managementserver": [
    {
      "id": "8d7b1557-cf29-44b6-8d35-876d1959d7f4",
      "name": "ghost",
      "state": "Up",
      "version": "4.16.0.0"
    }
  ]
}
(localcloud) 🐱 > list volumes virtualmachineid=8f411a8b-4fb5-4b23-9909-f6c9ccc62d0a filter=id,name,diskofferingid,type,
{
  "count": 2,
  "volume": [
    {
      "diskofferingid": "f32c64b7-22bd-4057-9456-e15773986455",
      "id": "8414beca-a378-48c5-bbaf-42bdbcff5dc7",
      "name": "DATA-3",
      "type": "DATADISK"
    },
    {
      "id": "14a335c2-ad48-4d33-abfd-ec9b45e9be20",
      "name": "ROOT-3",
      "type": "ROOT"
    }
  ]
}
(localcloud) 🐱 > list virtualmachines id=8f411a8b-4fb5-4b23-9909-f6c9ccc62d0a filter=id,name,diskofferingid,diskofferingname,
{
  "count": 1,
  "virtualmachine": [
    {
      "diskofferingid": "f32c64b7-22bd-4057-9456-e15773986455",
      "diskofferingname": "Medium",
      "id": "8f411a8b-4fb5-4b23-9909-f6c9ccc62d0a",
      "name": "VM-8f411a8b-4fb5-4b23-9909-f6c9ccc62d0a"
    }
  ]
}
(localcloud) 🐱 > list virtualmachines id=8f411a8b-4fb5-4b23-9909-f6c9ccc62d0a filter=id,name,diskofferingid,diskofferingname,templateid,templatename,
{
  "count": 1,
  "virtualmachine": [
    {
      "diskofferingid": "f32c64b7-22bd-4057-9456-e15773986455",
      "diskofferingname": "Medium",
      "id": "8f411a8b-4fb5-4b23-9909-f6c9ccc62d0a",
      "name": "VM-8f411a8b-4fb5-4b23-9909-f6c9ccc62d0a",
      "templateid": "cb8d1d7a-9370-11ee-9baf-9843fa67404e",
      "templatename": "CentOS 5.6 (64-bit) no GUI (Simulator)"
    }
  ]
}
```

**Comment by weizhouapache** [comment]
> @weizhouapache for ROOT disk and ISO we are already setting the diskoffering. This PR change will only affect with VM deployed from templates having multiple disks. In older version we were returning DATA disk's offering as disk offering
> 
> ```
> (localcloud) 🐱 > list managementservers 
> {
>   "count": 1,
>   "managementserver": [
>     {
>       "id": "8d7b1557-cf29-44b6-8d35-876d1959d7f4",
>       "name": "ghost",
>       "state": "Up",
>       "version": "4.16.0.0"
>     }
>   ]
> }
> (localcloud) 🐱 > list volumes virtualmachineid=8f411a8b-4fb5-4b23-9909-f6c9ccc62d0a filter=id,name,diskofferingid,type,
> {
>   "count": 2,
>   "volume": [
>     {
>       "diskofferingid": "f32c64b7-22bd-4057-9456-e15773986455",
>       "id": "8414beca-a378-48c5-bbaf-42bdbcff5dc7",
>       "name": "DATA-3",
>       "type": "DATADISK"
>     },
>     {
>       "id": "14a335c2-ad48-4d33-abfd-ec9b45e9be20",
>       "name": "ROOT-3",
>       "type": "ROOT"
>     }
>   ]
> }
> (localcloud) 🐱 > list virtualmachines id=8f411a8b-4fb5-4b23-9909-f6c9ccc62d0a filter=id,name,diskofferingid,diskofferingname,
> {
>   "count": 1,
>   "virtualmachine": [
>     {
>       "diskofferingid": "f32c64b7-22bd-4057-9456-e15773986455",
>       "diskofferingname": "Medium",
>       "id": "8f411a8b-4fb5-4b23-9909-f6c9ccc62d0a",
>       "name": "VM-8f411a8b-4fb5-4b23-9909-f6c9ccc62d0a"
>     }
>   ]
> }
> (localcloud) 🐱 > list virtualmachines id=8f411a8b-4fb5-4b23-9909-f6c9ccc62d0a filter=id,name,diskofferingid,diskofferingname,templateid,templatename,
> {
>   "count": 1,
>   "virtualmachine": [
>     {
>       "diskofferingid": "f32c64b7-22bd-4057-9456-e15773986455",
>       "diskofferingname": "Medium",
>       "id": "8f411a8b-4fb5-4b23-9909-f6c9ccc62d0a",
>       "name": "VM-8f411a8b-4fb5-4b23-9909-f6c9ccc62d0a",
>       "templateid": "cb8d1d7a-9370-11ee-9baf-9843fa67404e",
>       "templatename": "CentOS 5.6 (64-bit) no GUI (Simulator)"
>     }
>   ]
> }
> ```

@shwstppr 
thanks 
Let's move forward with your suggestion in https://github.com/apache/cloudstack/pull/8135#issuecomment-1832040649

@DaanHoogland what do you think ?

**Comment by DaanHoogland** [comment]
> @shwstppr thanks Let's move forward with your suggestion in [#8135 (comment)](https://github.com/apache/cloudstack/pull/8135#issuecomment-1832040649)

agreed :handshake:

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan test

**Review by weizhouapache — APPROVED** [review]
code lgtm
not tested it

**Pull request merged**